src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs

```diff
@@ -0,0 +1,336 @@
+
+                bool success = EffectiveConverter.WriteCore(writer, rootValue!, Options, ref state);
+                Debug.Assert(success);
+                writer.Flush();
+            }
+        }
+
+        // Root serialization method for async streaming serialization.
+        internal async Task SerializeAsync(
+            Stream utf8Json,
+            T? rootValue,
+            CancellationToken cancellationToken,
+            object? rootValueBoxed = null,
+            bool isInvokedByPolymorphicConverter = false)
+        {
+            Debug.Assert(IsConfigured);
+            Debug.Assert(rootValueBoxed is null || rootValueBoxed is T);
+
+            if (CanUseSerializeHandlerInStreaming)
+            {
+                // Short-circuit calls into SerializeHandler, if the `CanUseSerializeHandlerInStreaming` heuristic allows it.
+
+                Debug.Assert(SerializeHandler != null);
+                Debug.Assert(Options.SerializerContext?.CanUseSerializationLogic == true);
+                Debug.Assert(Converter is JsonMetadataServicesConverter<T>);
+
+                using var bufferWriter = new PooledByteBufferWriter(Options.DefaultBufferSize);
+                Utf8JsonWriter writer = Utf8JsonWriterCache.RentWriter(Options, bufferWriter);
+
+                try
+                {
+                    SerializeHandler(writer, rootValue!);
+                    writer.Flush();
+                }
+                finally
+                {
+                    OnRootLevelAsyncSerializationCompleted(writer.BytesCommitted + writer.BytesPending);
+                    Utf8JsonWriterCache.ReturnWriter(writer);
+                }
+
+                await bufferWriter.WriteToStreamAsync(utf8Json, cancellationToken).ConfigureAwait(false);
+            }
+            else if (
+#if NETCOREAPP
+                !typeof(T).IsValueType &&
+#endif
+                !isInvokedByPolymorphicConverter &&
+                Converter.CanBePolymorphic &&
+                rootValue is not null &&
+                Options.TryGetPolymorphicTypeInfoForRootType(rootValue, out JsonTypeInfo? derivedTypeInfo))
+            {
+                Debug.Assert(typeof(T) == typeof(object));
+                await derivedTypeInfo.SerializeAsObjectAsync(utf8Json, rootValue, cancellationToken, isInvokedByPolymorphicConverter: true).ConfigureAwait(false);
+            }
+            else
+            {
+                bool isFinalBlock;
+                WriteStack state = default;
+                state.Initialize(this,
+                    rootValueBoxed,
+                    isInvokedByPolymorphicConverter,
+                    supportContinuation: true,
+                    supportAsync: true);
+
+                using var bufferWriter = new PooledByteBufferWriter(Options.DefaultBufferSize);
+                using var writer = new Utf8JsonWriter(bufferWriter, Options.GetWriterOptions());
```

Comment:
Has any thought been put into making the cache use a `ConcurrentQueue` instead so it can be used in async paths?
Example: https://github.com/dotnet/aspnetcore/blob/main/src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketSenderPool.cs

Reply:
Given that the perf benefits are [fairly modest](https://github.com/dotnet/runtime/pull/73338) when doing thread-local caching in the sync methods, my expectation is that any mechanism requiring more elaborate synchronization would likely offer diminishing returns.

Reply:
The big gain we've seen in our JSON web benchmarks is that by reducing the allocations of the `Utf8JsonWriter` (aka pooling them) it greatly reduces the working set of the application, from over 400 MB to ~60 MB for a test run doing millions of RPS.

Reply:
> The big gain we've seen in our JSON web benchmarks is that by reducing the allocations of the `Utf8JsonWriter` (aka pooling them) it greatly reduces the working set of the application, from over 400 MB to ~60 MB for a test run doing millions of RPS.

Do we know how much of that would also be achieved by changing GC configuration to make it more aggressive rather than being ok expanding to the memory available?

Reply:
@BrennanConroy could we do a run of the JSON middleware benchmark with Workstation GC to compare?

Reply:
Server GC: 468MB working set
Workstation GC: 99MB working set

Server GC with manual caching: 86MB working set
Workstation GC with manual caching: 84MB working set

Reply:
Thanks @BrennanConroy. For the sake of curiosity, what were the RPS numbers for those runs?

The savings when pooling occurs seem worthwhile even when GC is more aggressive (15% for our JSON middleware benchmark).

Reply:
@Maoni0, if someone wants the scalability of server GC but also wants to prioritize minimizing working set, what are the right knobs to experiment with?

Reply:
None of the streaming methods accept `Utf8JsonWriter`, so I'm guessing these numbers came from benchmarking the [non-streaming method](https://learn.microsoft.com/en-us/dotnet/api/system.text.json.jsonserializer.serialize?view=net-7.0#system-text-json-jsonserializer-serialize-1(system-text-json-utf8jsonwriter-0-system-text-json-jsonserializeroptions)) to itself using caching? If not, I don't think it would be an apples-to-apples comparison, particularly from an allocation perspective.

Reply:
> I'm guessing these numbers came from benchmarking the [non-streaming method](https://learn.microsoft.com/en-us/dotnet/api/system.text.json.jsonserializer.serialize?view=net-7.0#system-text-json-jsonserializer-serialize-1(system-text-json-utf8jsonwriter-0-system-text-json-jsonserializeroptions)) to itself using caching?

Correct.

I also realized the numbers I shared before were wrong since we switched to using the source-generator at the same time.

So here are the numbers with just caching Utf8Writer, no source-gen or JsonConverter to avoid the struct boxing so there is ~24M bytes/sec allocations:

Server GC: 468MB working set
Workstation GC: 99MB working set

Server GC with manual caching: 427MB working set
Workstation GC with manual caching: 98MB working set

Basically, working set with server GC will be high if there are any allocations it looks like.

Reply:
This is maybe the only thing similar right now in the JSON code base https://github.com/dotnet/runtime/blob/b77aa8a9726503df52327a0388a3f4a0325989e1/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs#L166

Reply:
Could you share the benchmark code? If run in .NET 7, I would expect any perf issues related to `Utf8JsonWriter` allocations to have dissipated because of #73338.

Reply:
https://github.com/aspnet/Benchmarks/blob/main/src/Benchmarks/Middleware/JsonMiddleware.cs#L56
It's using the Stream methods which weren't affected until this PR.

Reply:
What about the variant that caches `Utf8JsonWriter` instances?

Reply:
Still streaming API, just the synchronous one
https://github.com/aspnet/Benchmarks/pull/1772